HttpProxy: Use CloseableHttpAsyncClient

matthiaso opened the pull request.

matthiaso committed:
HttpProxy: Use CloseableHttpAsyncClient

matthiaso commented on org.eclipse.scout.rt.server.commons/src/main/java/org/eclipse/scout/rt/server/commons/servlet/HttpProxy.java:
```diff
@@ -105,6 +121,12 @@ protected void initialize() {
+  }
+
+  protected ExecutorService createBlockingOperationExecutor() {
+    return Executors.newCachedThreadPool();
```
BEANS.get(JobManager.class).getExecutor()

matthiaso commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/http/async/BasicAuthAsyncHttpClientManager.java:
```diff
@@ -0,0 +1,35 @@
+  @Override
+  protected void interceptCreateClient(HttpAsyncClientBuilder builder) {
+    UsernamePasswordCredentials credentials = new UsernamePasswordCredentials(m_user, m_password.toCharArray());
+    builder.setDefaultCredentialsProvider((authScope, context) -> credentials);
```
That does not work the same way as org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/http/BasicAuthHttpTransportManager.java; previously always Authorization header was set preemptively for every request/host. - Further testing/fixing necessary.

paolobazzi commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/http/async/DefaultAsyncHttpClientManager.java:
```diff
@@ -0,0 +1,124 @@
+  protected void init() {
+    if (!m_initialized) {
+      initSynchronized();
+      m_initialized = true;
```
Set initialized as part of the synchronized block?

matthiaso replied:
Thought the same when I copied the code; moving it.

matthiaso replied:
Moved in 912450f97f0febfebac81326d3b7f89eb4bdca12

paolobazzi commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/http/async/DefaultAsyncHttpClientManager.java:
```diff
@@ -0,0 +1,124 @@
+    }
+    finally {
+      m_active = false;
+      m_client = null;
```
+ initialized=false

matthiaso replied:
Agree

matthiaso replied:
Added in 912450f97f0febfebac81326d3b7f89eb4bdca12

paolobazzi commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/http/async/DefaultAsyncHttpClientManager.java:
```diff
@@ -0,0 +1,124 @@
+        .create()
+        .useSystemProperties()
+        .build());
+    builder.useSystemProperties();
```
duplicated line

matthiaso replied:
Disagree, not duplicated - line 79 is called on a different builder.

paolobazzi commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/http/async/DefaultAsyncHttpClientManager.java:
```diff
@@ -0,0 +1,124 @@
+
+    builder.setConnectionManager(PoolingAsyncClientConnectionManagerBuilder
+        .create()
+        .useSystemProperties()
```
if global trust manager is setup correctly using system properties should not be necessery since system properties are loaded into global trust manager as fallback, org.eclipse.scout.rt.server.commons.GlobalTrustManager.P_GlobalTrustManager#P_GlobalTrustManager

matthiaso replied:
Yes, but w/o `useSystemProperties` the `javax.net.ssl.SSLContext.getDefault()` holding our trust manager is not used; in this case it would be necessary to specify trust manager explicitly here.

matthiaso replied:
Comment this method is used to use system trust manager.

Maybe only one call to useSystemProperties?

matthiaso replied:
Added comment and removed one call in 912450f97f0febfebac81326d3b7f89eb4bdca12

paolobazzi commented on org.eclipse.scout.rt.server.commons/src/main/java/org/eclipse/scout/rt/server/commons/servlet/AbstractHttpServlet.java:
```diff
@@ -48,12 +55,59 @@ protected void wrap(HttpServletRequest req, HttpServletResponse resp, HttpServle
       responseProxyHandler.invalidate();
     }
+    else {
+      AsyncContext asyncContext = req.getAsyncContext();
```
extract into separate invalidateAsync() method?

matthiaso replied:
Agree

matthiaso replied:
Extracted in 912450f97f0febfebac81326d3b7f89eb4bdca12

paolobazzi commented on org.eclipse.scout.rt.server.commons/src/main/java/org/eclipse/scout/rt/server/commons/servlet/AbstractHttpServlet.java:
```diff
@@ -48,12 +55,59 @@ protected void wrap(HttpServletRequest req, HttpServletResponse resp, HttpServle
+   * register a listener to call these methods on completion.
+   */
+  protected void invalidateRequestAndResponse(HttpServletRequest req, P_HttpInvocationHandler requestProxyHandler, P_HttpInvocationHandler responseProxyHandler) {
+    if (!req.isAsyncStarted()) {
```
change order:

```
if (req.isAsyncStarted()) {
  invalidateAsync();
} else {
...
}
```

matthiaso replied:
Agree, thanks for the input

matthiaso replied:
Changed in 912450f97f0febfebac81326d3b7f89eb4bdca12

paolobazzi commented on org.eclipse.scout.rt.server.commons/src/main/java/org/eclipse/scout/rt/server/commons/servlet/HttpProxy.java:
```diff
@@ -159,41 +186,48 @@ protected boolean shouldWriteParametersAsPayload(HttpServletRequest req) {
    * the options object. This allows to redirect the request to a different URL than the URL that has been requested.
    */
   protected String rewriteUrl(HttpServletRequest req, HttpProxyRequestOptions options) {
+    LOG.trace("Entering rewriteUrl");
```
Do we need this low level trace statements? (also others in this class)

matthiaso replied:
No, not really - was only used for me while debugging; however they may be useful if we really run into problems during further tests. In this case we could dynamically increase log level for this class. But on the other hand, other classes also do not provide such detailed tracing.

matthiaso replied:
Remove simple "entering method" logging outputs.

matthiaso replied:
Removed "entering method" outputs in 912450f97f0febfebac81326d3b7f89eb4bdca12

paolobazzi commented on org.eclipse.scout.rt.server.commons/src/main/java/org/eclipse/scout/rt/server/commons/servlet/HttpProxy.java:
```diff
@@ -404,4 +452,156 @@ public HttpProxy withResponseHeaderFilter(IHttpHeaderFilter filter) {
+        BEANS.get(ExceptionHandler.class).handle(ex);
+        boolean alreadyCommitted = resp.isCommitted();
+        if (!alreadyCommitted) {
+          resp.setStatus(500);
```
use HttpStatus.SC_INTERNAL_SERVER_ERROR

matthiaso replied:
Thanks

matthiaso replied:
Fixed in 912450f97f0febfebac81326d3b7f89eb4bdca12

paolobazzi commented on org.eclipse.scout.rt.server.commons/src/main/java/org/eclipse/scout/rt/server/commons/servlet/HttpProxy.java:
```diff
@@ -57,13 +75,16 @@
 
-  private IHttpTransportManager m_httpTransportManager;
+  private DefaultAsyncHttpClientManager m_clientManager;
+  private int m_initialBufferSize = 4096;
```
add and use getter for all private variables, or change to protected

matthiaso replied:
Thanks

matthiaso replied:
Added getters in 912450f97f0febfebac81326d3b7f89eb4bdca12

paolobazzi reviewed (commented):
See inputs

matthiaso commented:
See changes since last review in 

* 897daa2b3b6167a4228180779f6ac7aad3c6c652 and
* https://github.com/eclipse-scout/scout.rt/compare/912450f97f0febfebac81326d3b7f89eb4bdca12..80bf33fc12bc48f40813b98980212b5a06b13704

matthiaso commented on org.eclipse.scout.rt.server.commons.test/src/test/java/org/eclipse/scout/rt/server/commons/servlet/HttpProxyTest.java:
```diff
@@ -253,4 +307,169 @@ protected void assertRewriteResponseHeaders(Map<String, String> expectedHeaders,
+  @Test
+  public void testProxyRequest_largeContent() throws IOException {
+    // keep size pseudo-large (could be larger) on CI (at least as long as we do not run tests in parallel), however can be increased for testing locally
+    byte[] content = new byte[65536 * 4096];
```
Too much for CI? Test seems to fail on CI with OOME; either decrease size here or ignore it (only use it for local testing).

paolobazzi replied:
Suggest to decrease size.

matthiaso replied:
Fixed in https://github.com/eclipse-scout/scout.rt/compare/f48c418ea822284b166f9970e4124a60e383bd63..e779671a1c3dd3fdd03df96227b29f95f566207b

paolobazzi commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/http/async/AbstractAsyncHttpClientManager.java:
```diff
@@ -0,0 +1,117 @@
+public abstract class AbstractAsyncHttpClientManager<BUILDER> implements IPlatformListener {
+
+  /**
+   * Is {@link AbstractAsyncHttpClientManager} still active?
```
Javadoc without ?: Suggestion: Flag indicating if {@link ...} is still active.

matthiaso replied:
Fixed in https://github.com/eclipse-scout/scout.rt/compare/f48c418ea822284b166f9970e4124a60e383bd63..e779671a1c3dd3fdd03df96227b29f95f566207b

paolobazzi commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/http/async/AbstractAsyncHttpClientManager.java:
```diff
@@ -0,0 +1,117 @@
+  protected volatile boolean m_active = true;
+
+  /**
+   * Initialized?
```
Suggestion: Flag indicating if {@link... } was successfully initialized.

matthiaso replied:
Fixed in https://github.com/eclipse-scout/scout.rt/compare/f48c418ea822284b166f9970e4124a60e383bd63..e779671a1c3dd3fdd03df96227b29f95f566207b
